tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/api.py

```diff
@@ -260,29 +261,68 @@ class OperationError(Error):
 
-    def __init__(self, api_name, operation_response, message=None):
+    def __init__(self, api_name: str, response: dict):
         self.api_name = api_name
-        operation = json_format.ParseDict(
-            operation_response,
-            Operation(),
-            ignore_unknown_fields=True,
-            descriptor_pool=error_details_pb2.DESCRIPTOR.pool)
+
+        # Operation.metadata field is Any specific to the API. It may not be
+        # present in the default descriptor pool, and that's expected.
+        # To avoid json_format.ParseError, handle it separately.
+        self.metadata = response.pop('metadata', {})
+
+        # Must be after removing metadata field.
+        operation: Operation = self._parse_operation_response(response)
         self.name = operation.name or 'unknown'
         self.code_name = code_pb2.Code.Name(operation.error.code)
         self.error = operation.error
-        # Collect error details packed as Any without parsing concrete types.
-        self.error_details = [
-            text_format.MessageToString(any_error, as_one_line=True)
-            for any_error in self.error.details
-        ]
-        if message is None:
-            message = (f'{api_name} operation "{self.name}" failed. Error '
-                       f'code: {self.error.code} ({self.code_name}), '
-                       f'message: {self.error.message}, '
-                       f'details: {self.error_details}')
-        self.message = message
-        super().__init__(message)
+        super().__init__()
+
+    @staticmethod
+    def _parse_operation_response(operation_response: dict) -> Operation:
+        try:
+            return json_format.ParseDict(
+                operation_response,
+                Operation(),
+                ignore_unknown_fields=True,
+                descriptor_pool=error_details_pb2.DESCRIPTOR.pool)
+        except (json_format.Error, TypeError) as e:
+            # Swallow parsing errors if any. Building correct OperationError()
+            # is more important than losing debug information. Details still
+            # can be extracted from the warning.
+            logger.warning(
+                ("Can't parse response while processing OperationError: '%r', "
+                 "error %r"), operation_response, e)
+            return Operation()
+
+    def __str__(self):
+        indent_l1 = ' ' * 2
+        indent_l2 = indent_l1 * 2
+
+        result = (f'{self.api_name} operation "{self.name}" failed.\n'
+                  f'{indent_l1}code: {self.error.code} ({self.code_name})\n'
+                  f'{indent_l1}message: "{self.error.message}"')
+
+        if self.error.details:
+            result += f'\n{indent_l1}details: [\n'
+            for any_error in self.error.details:
+                error_str = json_format.MessageToJson(any_error)
+                for line in error_str.splitlines():
+                    result += indent_l2 + line + '\n'
+            result += f'{indent_l1}]'
```

Comment:
`result += f'{indent_l1}]\n'`?

Reply:
Sorry, the space should be in the loop of `for any_error in self.error.details:`. Using a pseudo code snippet:

```py
for any_error in self.error.details:
    for line in _:
        result += ...
    result += '\n'  # <-
result += "]\n"
```

Reply:
Not sure we need the new line at the end of the message. In fact when printing metadata below I intentionally strip it out: `result = result.rstrip()`. I think python prints exception messages with the new line at the end anyway, so I was removing an extra new line. But I could be wrong.

Reply:
My comment above was a response to the first one in this thread, before I saw you follow up. Let me see what you mean.

Reply:
My bad, you are right, we should not add a trailing line break. I'm commenting about separate different errors in `self.error.details`.

Reply:
I just checked the output of my code with two error details, it looks correct:

```
framework.infrastructure.gcp.api.OperationError: networksecurity operation "projects/grpc-testing/locations/global/operations/operation-1629948909762-5ca6e0dc4b74d-a93ac167-5885c4ed" failed.
  code: 3 (INVALID_ARGUMENT)
  message: "Config validation failed"
  details: [
    {
      "@type": "type.googleapis.com/google.rpc.BadRequest",
      "fieldViolations": [
        {
          "field": "resource.name1",
          "description": "line 1"
        }
      ]
    }
    {
      "@type": "type.googleapis.com/google.rpc.BadRequest",
      "fieldViolations": [
        {
          "field": "resource.name2",
          "description": "line 2"
        }
      ]
    }
  ]
```

Reply:
We could add a comma between the items, but I wasn't going for a full JSON correctness.

Reply:
Ah, `Any` message has outer parenthesis. Thanks for checking this!

Reply:
Ah, I see what you mean now. Thanks you for considering this case!